File: crates/pattern_runtime/haskell/Pattern/Display.hs
{-# LANGUAGE GADTs #-}
-- | Pattern.Display — one-way broadcast of observable agent output to
-- UX / CLI / telemetry surfaces.
--
-- Fully implemented in Phase 3 (broadcast to registered subscribers).
--
-- From the agent's perspective this is fire-and-forget: the agent emits
-- `Chunk` / `Final` / `Note` envelopes describing what the human / UX
-- layer should see. Subscribers are registered host-side; see the
-- `DisplayHandler` in @pattern_runtime::sdk::handlers::display@.
--
-- Typical flow: the runtime handler `MessageHandler` forwards streaming provider
-- chunks through `Display.Chunk` in real time, then emits `Display.Final`
-- with the assembled content. Agent Haskell programs that want to react
-- mid-stream should register a Display subscriber rather than attempting
-- to iterate chunks in Haskell.
module Pattern.Display where

import Control.Monad.Freer (Eff, Member, send)
import Data.Text (Text)

-- | Effect algebra.
data Display a where
  -- | Incremental chunk during a streaming provider response.
  Chunk :: Text -> Display ()
  -- | Terminal assembled content for the turn's Message.Ask. Fires once.
  Final :: Text -> Display ()
  -- | Agent-visible note (typing indicator, tool-call progress, etc.).
  Note  :: Text -> Display ()

chunk :: Member Display effs => Text -> Eff effs ()
chunk t = send (Chunk t)

final :: Member Display effs => Text -> Eff effs ()
final t = send (Final t)

note :: Member Display effs => Text -> Eff effs ()
note t = send (Note t)

File: crates/pattern_runtime/src/sdk/requests/display.rs
// Copyright 2026 Pattern contributors
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, you can obtain one at http://mozilla.org/MPL/2.0/.

//! Mirror of `Pattern.Display` (`haskell/Pattern/Display.hs`).
//!
//! The Display effect is broadcast-style: the Haskell agent emits one-shot
//! envelopes describing observable output. Subscribers registered Rust-side
//! receive them in realtime. See `sdk/handlers/display.rs` for subscriber
//! shape.

use tidepool_bridge_derive::FromCore;

/// Rust mirror of the Haskell `Display` GADT.
#[derive(Debug, FromCore)]
pub enum DisplayReq {
    /// A partial chunk during a streaming provider response. Forwarded to
    /// every registered subscriber as-is.
    #[core(module = "Pattern.Display", name = "Chunk")]
    Chunk(String),
    /// Final assembled content for the turn's Message.Ask. Fires once,
    /// after the provider stream completes.
    #[core(module = "Pattern.Display", name = "Final")]
    Final(String),
    /// Agent-visible note (typing indicator, tool-call progress, etc.) that
    /// isn't part of the LLM response stream. Subscribers decide whether
    /// to render.
    #[core(module = "Pattern.Display", name = "Note")]
    Note(String),
}
